#33 1 implement edit module

PhilippeLeopoldie opened the pull request:
This PR add module creation and editing functionality

- Introduced ModuleForm.razor, a reusable form for creating and editing modules with validation and improved layout.

- Integrated ModuleForm into CreateModule.razor and added a new EditModule.razor page with teacher authorization and API update handling.

- Enhanced module display in ModulesComponent.razor by showing start/end dates and improved layout via CourseRow with customizable menus.

- Enforced required Description in DTOs for better data integrity.

- Added Result<T> utility class for operation outcomes and introduced ModuleStateService for scoped state management.

- Refactored ModuleComponent and supporting logic for consistency, improved navigation, and cleaner structure.

PhilippeLeopoldie committed:
Add module creation/editing form in ModuleForm.razor

This update introduces a new form for creating or editing modules. It features an `EditForm` component bound to a `FormModel`, with fields for the module's name, description, start date and time, and end date and time, each accompanied by validation messages. A validation summary and a submit button are included, along with parameters for the `FormModel` and an event callback for form submission.

PhilippeLeopoldie committed:
Replace EditForm with ModuleForm in CreateModule.razor

Updated the form implementation by replacing the `<EditForm>` component with a new `<ModuleForm>` component, maintaining the same `FormModel` and `OnValidSubmit` parameters. Retained error message display for better user feedback, while keeping the access control within the `<AuthorizeView>` for "Teacher" role users.

PhilippeLeopoldie committed:
Update module display to include start and end dates

The `CourseRow` component in `ModulesComponent.razor` now shows the module's start and end dates alongside its name. This enhances the information presented to users, while the activity count remains unchanged.

PhilippeLeopoldie committed:
Enforce required Description in DTOs

Updated the `Description` property in `ModuleDto` and `ModuleForManipulationDto` from nullable to required. Added validation to ensure that a description must be provided, enhancing data integrity for these data transfer objects.

PhilippeLeopoldie committed:
Add generic Result<T> class for operation outcomes

Introduced a new generic class `Result<T>` in the
`LMS.Blazor.Client.Pages.Modules` namespace. This class
provides properties for operation success, result data,
and error messages. It includes a private constructor
and two static methods: `Ok(T data)` for successful
results and `Fail(string errorMessage)` for failures.

PhilippeLeopoldie committed:
Refactor ModuleComponent for improved UI and functionality

Renamed `courseId` to `CourseId` for consistency.
Introduced `ModuleMenu` method for dropdown actions.
Restructured UI components for better layout control.
Updated API calls to use the new `CourseId` property.

PhilippeLeopoldie committed:
Add ModuleStateService for module state management

Introduces `ModuleStateService` to the dependency injection
container in `Program.cs`, allowing for scoped management of
module state within the Blazor client. A new file,
`ModuleStateService.cs`, defines the service and its
`Module` property of type `ModuleDto?`. Existing service
registrations for API and authentication have been modified
to reflect this change.

PhilippeLeopoldie committed:
Add CourseId parameter to ModuleForm and refactor submit

Updated the `ModuleForm` in `CreateModule.razor` to include a new `CourseId` parameter for better context. Refactored the `HandleValidSubmit` method to improve clarity by moving navigation logic to a new `NavigateBack` method. Error handling remains unchanged to ensure proper exception management.

PhilippeLeopoldie committed:
Replace ModulesComponent with ModuleComponent

Updated the Razor component to replace the `ModulesComponent` with a `ModuleComponent`, indicating a shift to focus on a single module representation. The surrounding context and header actions remain unchanged.

PhilippeLeopoldie committed:
Enhance ModuleForm with improved structure and navigation

Updated `ModuleForm.razor` to use an `EditForm` for better validation handling. Restructured labels and input fields, adding `SlutDatum` (End Date) and `Beskrivning` (Description). Improved layout with Bootstrap classes and added a "Cancel" button for navigation. Introduced a new `CourseId` parameter and a `NavigateBack` method for easier navigation to course modules.

PhilippeLeopoldie committed:
Add customizable menu content to CourseRow component

Introduced a new `MenuContent` parameter of type `RenderFragment?` to the `CourseRow` component for customizable menu display. Conditional rendering is implemented to show `MenuContent` in a `div` if not null, or a button to trigger the menu if `MenuContent` is null but `HasMenu` is true.

PhilippeLeopoldie committed:
Add EditModule page with authorization and form handling

Updated `EditModule.razor` to create a new page for editing modules within a course. Added authorization for "Teacher" role, a form for editing module details, and error handling for submissions. Initialized form model with existing module data and implemented API update requests. Removed unused method for fetching modules by course ID.

PhilippeLeopoldie committed:
EditModule: Update header with authorization check

Replaced the static header with an `<AuthorizeView>` component to ensure only users with the "Teacher" role can see the module name.

PhilippeLeopoldie committed:
Merge branch 'development' into #33-1-implement-EditModule

JorgenHagglund reviewed (approved):
Ser bra ut.  Funderar bara på property'n HasMenu i CourseRow.razor, behövs den?

PhilippeLeopoldie committed:
Remove ActionName attribute from GetCourseByIdAsync

The `ActionName` attribute for the `GetCourseByIdAsync` method was removed from the `CoursesController` class. This change may affect how the action is referenced in routes or when generating links, as the explicit action name is no longer defined.

ahlgren1234 reviewed (approved):
Ser bra ut.

ChornaIryna commented on LMS.Blazor.Client/Pages/Modules/Result.cs:
Result<T> är en bra idé, men används inte. kommer du använda den eller glömde ta bort?

PhilippeLeopoldie replied:
Ja, jag kommer att använda det, jag ville bara inte ändra för mycket innan presentationen.

ChornaIryna commented on LMS.Blazor.Client/Pages/Modules/ModuleComponent.razor:
```diff
@@ -47,18 +53,50 @@ else
 
+                <!-- Delete -->
+                <button class="dropdown-item btn btn-link text-danger" @onclick="() => Delete()">
+                    <i class="bi bi-trash"></i> Ta bort
```
Behövs den Delete knappen? Delete() är en placeholder som bara navigerar – ingen faktisk borttagning sker.

PhilippeLeopoldie replied:
Ja, jag kommer att använda det, men jag ville bara inte ändra för mycket innan presentationen. Därför har jag kommenterat i Delete-metoden att det bara var en temporär route.

ChornaIryna commented on LMS.Blazor.Client/Pages/Modules/EditModule.razor:
```diff
@@ -0,0 +1,93 @@
+        }
+    }
+
+    /*private async Task<Result<ModuleDto?>> GetModuleByCourseId(int id)
```
behövs den kommenterad kod?

PhilippeLeopoldie replied:
Ja, jag ville spara ett 'Result<T>' exempel som jag kommer att använda efter presentationen.

ChornaIryna replied:
Ok, då vet jag. Bra jobbat!

ChornaIryna commented on LMS.Blazor.Client/Pages/Modules/EditModule.razor:
```diff
@@ -0,0 +1,93 @@
+@page "/courses/{courseId:int}/module/{moduleId:int}/edit"
+
+
+
```
Du kan lägga till spinner eller annat laddningsindikator här

ChornaIryna reviewed (approved):
👍🏻
